packages/dds/merge-tree/src/partialLengths.ts

```diff
@@ -14,35 +14,106 @@ import {
     ISegment,
     toRemovalInfo,
 } from "./mergeTreeNodes";
+import { SortedSet } from "./sortedSet";
 
-interface IOverlapClient {
-    clientId: number;
-    seglen: number;
-}
+class PartialSequenceLengthsSet extends SortedSet<PartialSequenceLength, number> {
+    protected getKey(item: PartialSequenceLength): number {
+        return item.seq;
+    }
 
-/**
- * Returns the partial length whose sequence number is
- * the greatest sequence number within a that is
- * less than or equal to key.
- * @param a - array of partial segment lengths
- * @param key - sequence number
- */
-function latestLEQ(a: PartialSequenceLength[], key: number) {
-    let best = -1;
-    let lo = 0;
-    let hi = a.length - 1;
-    while (lo <= hi) {
-        const mid = lo + Math.floor((hi - lo) / 2);
-        if (a[mid].seq <= key) {
-            if ((best < 0) || (a[best].seq < a[mid].seq)) {
-                best = mid;
+    public addOrUpdate(
+        newItem: PartialSequenceLength,
+        update?: (existingItem: PartialSequenceLength, newItem: PartialSequenceLength) => void,
+    ) {
+        const prev = this.latestLEQ(newItem.seq);
+
+        if (prev?.seq !== newItem.seq) {
+            // new element, update len
+            newItem.len = (prev?.len ?? 0) + newItem.seglen;
+        }
+
+        // update the len of all following elements
+        for (const e of this.keySortedItems) {
+            if (e.seq <= newItem.seq) {
+                continue;
+            }
+
+            e.len += newItem.seglen;
+        }
+
+        super.addOrUpdate(newItem, (currentPartial, partialLength) => {
+            currentPartial.seglen += partialLength.seglen;
+            currentPartial.len += partialLength.seglen;
+            combineOverlapClients(currentPartial, partialLength);
+        });
+    }
+
+    /**
+     * Returns the partial length whose sequence number is the greatest sequence
+     * number that is less than or equal to key.
+     * @param key - sequence number
+     */
+    latestLEQ(key: number): PartialSequenceLength | undefined {
```

Comment:
I would have expected to see methods analogous to this on the base class rather than here. I'd also make the spelling consistent. Possible alternative names to `latestLEQ` and `firstGTE` are `floor` and `ceil` in that order.

Reply:
ideally we want to dedupe the binary search logic. Potentially might require changing how we store the segments (right now they're a flat list even if there are duped keys)